src/VisualStudio/IntegrationTest/TestUtilities/IntegrationHelper.cs

```diff
@@ -202,172 +182,49 @@ public static void SetForegroundWindow(IntPtr window)
             NativeMethods.SwitchToThisWindow(activeWindow, true);
         }
 
-        public static void SendInput(NativeMethods.INPUT[] inputs)
+        public static void SendInput(KeyPress[] inputs)
         {
             // NOTE: This assumes that Visual Studio is the active foreground window.
+            var simulator = new InputSimulator();
 
-            LogKeyboardInputs(inputs);
-
-            var eventsInserted = NativeMethods.SendInput((uint)inputs.Length, inputs, NativeMethods.SizeOf_INPUT);
-
-            if (eventsInserted == 0)
-            {
-                var hresult = Marshal.GetHRForLastWin32Error();
-                throw new ExternalException("Sending input failed because input was blocked by another thread.", hresult);
-            }
-        }
+            ResetKeyState(VirtualKeyCode.LCONTROL);
+            ResetKeyState(VirtualKeyCode.RCONTROL);
+            ResetKeyState(VirtualKeyCode.LSHIFT);
+            ResetKeyState(VirtualKeyCode.RSHIFT);
+            ResetKeyState(VirtualKeyCode.LMENU);
+            ResetKeyState(VirtualKeyCode.RMENU);
 
-        [Conditional("DEBUG")]
-        private static void LogKeyboardInputs(NativeMethods.INPUT[] inputs)
-        {
             foreach (var input in inputs)
             {
-                switch (input.Type)
+                if (!input.IsTextEntry && input.VirtualKey == VirtualKeyCode.ESCAPE)
                 {
-                    case NativeMethods.INPUT_KEYBOARD:
-                        LogKeyboardInput(input.ki);
-                        break;
-                    case NativeMethods.INPUT_MOUSE:
-                        Debug.WriteLine("UNEXPECTED: Encountered mouse input");
-                        break;
-                    case NativeMethods.INPUT_HARDWARE:
-                        Debug.WriteLine("UNEXPECTED: Encountered hardware input");
-                        break;
-                    default:
-                        Debug.WriteLine($"ERROR: Encountered illegal input type: {input.Type}");
-                        break;
+                    // Slight delay on Esc to ensure the Ctrl key isn't registered as down
```

Comment:
Better to spin until it's not? That'll doubly help if it turns out there's a deeper issue, because then we'll see hangs in a place where we'll understand what's up.

Reply:
(Also maybe link to a tracking bug? Since it's obvious to me why Escape is more special than any other modifier here. Yes, it has the added affect of launching explorer, but Ctrl+any other letter may invoke random VS commands which will confuse us just as much.)

Reply:
Ctrl+Esc is handled by the DWM, not Visual Studio. I couldn't find any evidence that the Ctrl key state is latching inside Visual Studio.

Reply:
I was not able to reproduce this locally even with high-speed uninterrupted toggling of Ctrl and Esc in thousands of iterations, so I haven't found any reason to believe a specific spin mechanism would be effective in the scenario.